node/authorship/src/tests.rs

```diff
@@ -702,57 +677,705 @@ fn terminal_extrinsic_discarded_from_txpool() {
+        genesis_hash.into(),
+        0_u32,
+        false.into(),
+        pre_digest(1, 0),
+        backend.as_ref(),
+    )
+    .unwrap();
+
+    extrinsics.into_iter().for_each(|xt: OpaqueExtrinsic| {
+        assert_ok!(block_builder.push(xt));
+    });
 
-    let timestamp_provider = sp_timestamp::InherentDataProvider::from_system_time();
+    assert_eq!(block_builder.extrinsics().len(), 5);
 
-    let proposer = block_on(
-        proposer_factory.init(
-            &client
-                .header(new_header)
-                .expect("Database error querying block #0")
-                .expect("Block #0 should exist"),
-        ),
+    // At this point the overlay wrapped in the `Api` instance has some changes
+    let fresh_block_builder = BlockBuilder::new(
+        client.as_ref(),
+        genesis_hash.into(),
+        0_u32,
+        false.into(),
+        pre_digest(1, 0),
+        backend.as_ref(),
     )
-    .expect("Proposer initialization failed");
+    .unwrap();
+
+    let cloned_block_builder = block_builder.clone();
 
-    let inherent_data =
-        block_on(timestamp_provider.create_inherent_data()).expect("Create inherent data failed");
-    let time_slot = sp_timestamp::Timestamp::current().as_millis() / SLOT_DURATION;
+    let (ext_1, api_1, ver_1, phash_1, bd_1, hsize_1) = block_builder.deconstruct();
+    let (ext_2, api_2, ver_2, phash_2, bd_2, hsize_2) = cloned_block_builder.deconstruct();
 
-    let proposal = block_on(proposer.propose(
-        inherent_data,
-        pre_digest(time_slot, 0),
-        time::Duration::from_secs(600),
+    // Assert that the components are equal but different
+    assert_eq!(ext_1, ext_2);
+    assert_ne!(ext_1.as_ptr(), ext_2.as_ptr());
+    let api_1_ptr: *const <RA as sp_api::ConstructRuntimeApi<_, Client<_, _, _, RA>>>::RuntimeApi =
+        api_1.deref();
+    let api_2_ptr: *const <RA as sp_api::ConstructRuntimeApi<_, Client<_, _, _, RA>>>::RuntimeApi =
+        api_2.deref();
+    assert_ne!(api_1_ptr, api_2_ptr);
+
+    // Reconstruct original block builders
+    let block_builder = BlockBuilder::<'_, _, Client<_, _, _, RA>, _>::from_parts(
+        ext_1, api_1, ver_1, phash_1, bd_1, hsize_1,
+    );
+    let cloned_block_builder = BlockBuilder::<'_, _, Client<_, _, _, RA>, _>::from_parts(
+        ext_2, api_2, ver_2, phash_2, bd_2, hsize_2,
+    );
+
+    let changes_1 = block_builder.into_storage_changes().unwrap();
+    let changes_2 = cloned_block_builder.into_storage_changes().unwrap();
+    let changes_3 = fresh_block_builder.into_storage_changes().unwrap();
+
+    // Assert that the original and the cloned block builders produce same storage changes
+    assert_eq!(
+        changes_1.transaction_storage_root,
+        changes_2.transaction_storage_root
+    );
+    // that are different from what builder created from scratch produces
+    assert_ne!(
+        changes_1.transaction_storage_root,
+        changes_3.transaction_storage_root
+    );
```

Comment:
Construct back and assert eq

Reply:
Actually, the last part of this test is exactly about this. Since `BlockBuilder` instances (both reconstructed from parts) can't be tested for equality directly (not being `Eq`), we compare their derivatives - storage changes.
Or do you mean anything else?